### PR TITLE
add in-memory secret cache for NodeAgent

### PR DESCRIPTION
**Pull request by quanjielin**
https://github.com/istio/istio/issues/5903

This PR includes:
1. in-memory secret cache
2. key rotation job

To be added in later PRs:
1. thread secret cache into SDS
2. move caClient to separated pkg and add implementation

### CODE REVIEW

File: security/pkg/nodeagent/cache/secretcache.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	sc.secrets.Store(proxyID, *ns)
+	return ns, nil
+}
+
+// Close shuts down the secret cache.
+func (sc *SecretCache) Close() {
+	sc.closing <- true
+}
+
+func (sc *SecretCache) keyRotationJob() {
```
Comment (incfly) [inline security/pkg/nodeagent/cache/secretcache.go]: this will be certificate rotation?

File: security/pkg/nodeagent/cache/secretcache.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/security/pkg/pki/util"
+)
+
+// The size of a private key for a leaf certificate.
+const keySize = 2048
+
+// CAClient interface defines the clients need to implement to talk to CA for CSR.
+// TODO(quanlin): CAClient here is a placeholder, will move it to separated pkg.
+type CAClient interface {
```
Comment (incfly) [inline security/pkg/nodeagent/cache/secretcache.go]: Here we actually assemble and sending CSR, not performing signing logic. Can we rename it to something more accurate? Like `FetchKeyCert` or something else?

https://github.com/istio/istio/blob/master/security/pkg/pki/ca/ca.go#L59 want to distinguish with `Sign` here.

File: security/pkg/nodeagent/cache/secretcache.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+			}()
+		}
+
+		return true
+	})
+}
+
+func (sc *SecretCache) generateSecret(jwtToken string, t time.Time) (*SecretItem, error) {
+	options := util.CertOptions{
+		Host:       jwtToken,
```
Comment (incfly) [inline security/pkg/nodeagent/cache/secretcache.go]: jwtToken is a credential should not be the value of `Host`.

Since it's just a prototype, it's fine to leave it empty & todo.

Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: I remember Citadel uses service account as certoption.host, not sure if it applies here until integration story is clear, my understanding is as long as it's a unique value should be fine ? anyway, remove jwtToken and add TODO.

File: security/pkg/nodeagent/cache/secretcache.go (hunk 2 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: cert is generated locally, so fetch** sounds a bit confusing to me. The caClient implementation need to assemble CSR then send request to corresponding CA, I feel at abstract level 'sign' is fine(interface hides implementation details of signing locally or remotely)

**Comment by quanjielin** [comment]
I don't think the test failure is related to this PR, we may want to remove some 'required' pre-checkin(pilot, mixer test) for collab-gcp-identity, where only security folder is touched. @sebastienvas

File: security/pkg/nodeagent/cache/secretcache.go (hunk 2 above)
Comment (incfly) [reply security/pkg/nodeagent/cache/secretcache.go]: To be precise, cert is generated locally but it's the Google CA signed that.

Just looking it here, I agree with your point that `Sign` can treated as an abstract operation. But if we look at other places,
- `pkg/caclient` sends request 
- In `pkg/pki/ca/ca.go`, the Sign means just a locally Sign, no CSR involved.
- Here the `Sign` means both.

Even we develop separate libraries, I hope we can be consistent of the terminology otherwise it'll very hard to tack & understand the code.

@myidpt

File: security/pkg/nodeagent/cache/secretcache.go (hunk 3 above)
Comment (incfly) [reply security/pkg/nodeagent/cache/secretcache.go]: Citadel is using SPIFFE id for the host, which includes the k8s service account.
https://github.com/istio/istio/blob/33f6c791bf44f4814ed34815b2ebe61020fb574d/security/pkg/pki/ca/controller/secret.go#L297

We need to make sure the `CertOptions.Host` and x509.SAN field represents the actual identity. In Google world, that corresponds to a GCP service account. Token is just to get CSR approved.

File: security/pkg/nodeagent/cache/secretcache.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+	return &SecretItem{
+		CertificateChain: certChainPER,
+		PrivateKey:       keyPEM,
+		jwtToken:         jwtToken,
+		lastUsedTime:     t,
+		createdTime:      t,
+	}, nil
+}
+
+func (sc *SecretCache) checkExpired(s *SecretItem) bool {
```
Comment (incfly) [inline security/pkg/nodeagent/cache/secretcache.go]: I think there're some complexities in rotating.

- First, if node agent needs to proactively rotate key cert, it should refresh key cert before TTL reaches, instead of `now.After(...)`.
- Node agent is not always able to send CSR to Google CA and push the results to envoy. If the metadata credential expires already (which is only updated when a pull SDS request comes from envoy), node agent can't impersonate workload.
- Envoy can configure the SDS request frequency like every x seconds. We can make that frequency smaller than the cert 
- So to make node agent proactive refreshing meaningful, you need to make OAuth2 token valid between two SDS request and the key cert TTL is shorter than the SDS request interval.

In summary, for short term we can rely on Envoy's configuration. 

@JimmyCYJ  @mandarjog Can you comment on this?

File: security/pkg/nodeagent/cache/secretcache.go [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+		closing:          make(chan bool, 1),
+	}
+
+	atomic.StoreUint64(&ret.secretChangedCount, 0)
+	go ret.keyCertRotationJob()
+	return ret
+}
+
+// GetSecret gets secret from cache.
+func (sc *SecretCache) GetSecret(proxyID, token string) (*SecretItem, error) {
```
Comment (incfly) [inline security/pkg/nodeagent/cache/secretcache.go]: Do you want  to use proxyID(assuming it means the envoy instance) as the key? How is that information obtained by node agent?  Can we use the identity itself as the key? Which seems a bit more straightforward to me, since the key/certs are bound with an identity, more than an instance.

File: security/pkg/nodeagent/cache/secretcache.go (hunk 4 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: Key rotation here is only for SDS.streamSecret. proxy periodically calls SDS.fetchSecret with interval set in envoy configuration; while for streamSecret, client only calls it once to setup connection, all later update is pushed from SDS to envoy, proactive push is one advantage of envoy v2 api to support grpc.

@qiwzhang, @mangchiandjjoe

File: security/pkg/nodeagent/cache/secretcache.go (hunk 2 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: renamed to CSRSign.

File: security/pkg/nodeagent/cache/secretcache.go (hunk 3 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: I see, thanks !

File: security/pkg/nodeagent/cache/secretcache.go (hunk 4 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: per decision from last Friday 
updated on server side - 
1. re-generate key/cert for every fetchsecret call.
2. keep current cache/key rotation for streamsecret.

client side(@JimmyCYJ)-
proxy uses fetch instead of stream for now, we could tweak this part later once having better idea of how often token expire, etc.

File: security/pkg/nodeagent/cache/secretcache.go (hunk 5 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: discovery request passing to SDS contains proxyID, I think it's OK for it in this initial PR and we could tweak it later once having better integration picture.

**Comment by sebastienvas** [comment]
please create branch from your fork.
